### PR TITLE
Fixes #11458 - Override inline for class parameters

### PR DESCRIPTION
**Pull request by tbrisker**
This implements override inline for smart class parameters and smart
variables in host and hostgroup edit forms.
Global variable override inline will be done in a seperate PR.

Authors:
    Ori Rabin orrabin@redhat.com
    Tomer Brisker tbrisker@redhat.com

### CODE REVIEW

File: app/helpers/lookup_keys_helper.rb [hunk 1]
```diff
@@ -74,35 +74,39 @@ def overridable_lookup_keys(klass, host)
...
+        diagnostic_class = 'has-error'
         diagnostic_helper = popover('', _("Required parameter without value.<br/><b>Please override!</b> <br><br><b>Description:</b>: %s") % key.description, :icon => "exclamation-sign")
       else
-        diagnostic_class = 'warning'
+        diagnostic_class = 'has-warning'
         diagnostic_helper = popover('', _("Optional parameter without value.<br/><i>Won\'t be given to Puppet.</i> <br><br><b>Description:</b> %s") % key.description, :icon => "warning-sign")
       end
     else
-      diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => matcher}, :data => { :placement => 'top' })
+      diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br/><b>Type:</b> %{type}<br/> <b>Matcher:</b> %{matcher} <br/><b>Value:</b> %{original_value}") % { :desc => key.description, :type => key.key_type, :matcher => matcher, :original_value => original_value }, :data => { :placement => 'top' }, :title => _("Original value info"))
```
Comment (domcleal) [inline app/helpers/lookup_keys_helper.rb]: Is it worth calling this the inherited or original value, rather than value?

**Comment by domcleal** [comment]
1) When editing a host, global parameters inherited from a host group etc. no longer are readonly - they have editable text boxes.

2) Moving the use_puppet_default help text to the top instead of next to every checkbox is good, but consider adding a tooltip to the checkbox just so you know what it is.  It doesn't need the full explanation, just the title.

3) The value fields are a lot more narrow than the previous overridden value fields, and slightly narrower than the inherited value fields used to be.  I think this will be OK because of the full screen button, but can you see if we can reclaim any more horizontal space?

Perhaps from the Use Puppet default column or even going back to icons instead of override/remove buttons?

4) It would be good to highlight overridden fields more, as this was a benefit for some users of the old layout.  The background colour of the text field and presence of the checkbox are the main visual clues right now, but I'd suggest also changing the border colour of the value text boxes (perhaps to blue?) or making the parameter name bold.

Else probably something for a future PR, some links/buttons at the top to filter/only view overridden parameters.

5) Can the remove/override buttons be the same width?

6) I don't think the PR makes it any worse than it currently is, but FYI I'm bumping into this nasty issue when overriding YAML lookup keys inherited from host groups: http://projects.theforeman.org/issues/12284

Looks really good, I'm impressed!

File: app/helpers/lookup_keys_helper.rb (hunk 1 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: Note that the popover title states this is all info related to the original value, do you think this needs more clarificaion?

**Comment by tbrisker** [comment]
@domcleal I have updated to original commit to address your first comment, and added the corrections for the rest of the comments to a separate commit which can be squashed later if accepted, just to allow simple reversion if we decide to leave it as is and address those issues in the future.

**Comment by dLobatog** [comment]
@tbrisker I've taken a look - not properly reviewed it at all, but from the UI/UX point of view.
- Every time I select an environment, a new header Puppet class parameters is added. See it in action 

demo gif parameters duplicate environment
- The yellow background color for overrides is doesn't fit very well I would say. Also for parameters without value, it shows the warning in yellow by default (when it's not overridden), and the border in red?

yellow override. 

What @domcleal suggested seems to look better, I would also suggest to display a small popover 'Overridden' when you hover the mouse on that table row:

screenshot from 2015-11-05 13-39-50
- The override button looks very out of place, and it shows up on the Name cell. Why not adding it to the value textarea? Or even the actions column is better than this one I'd say.
  
  override button
- Long Puppet class names are truncated but it'd be nice to see the whole name on mouse hover.
- 'Use puppet default' doesn't work at all? It doesn't show checkboxes.

Let me know what you think :ok_hand:

**Comment by tbrisker** [comment]
@dLobatog I addressed most of your comments. 
For the values that have an empty default - if they are not required they show a yellow warning and the border is yellow, if they are required the border is red. I have changed it now so that clicking on override will remove the border.
Long class names should show whole name on hover, however the tooltip is attached to the `<td>` so it might appear higher if the class has multiple parameters. 
'Use puppet default' checkbox is only displayed for smart class parameters, not smart variables as they have no puppet default.

**Comment by dLobatog** [comment]
@tbrisker Thanks! A few more:
- I don't see any tooltips on hover
- Top left and bottom left border-radius of the `override` button should be 0px to blend in with the full screen button.
- Is this the way required empty values are supposed to look? It looks a bit strange in my opinion, the warning sign is enough in my opinion. I can't see any guidelines for highlighting elements in a table on [Patternfly](patternfly.org/) @kybaker any hints?
   

This doesn't have a lot to do with the PR, but what do you think about softening the borders of the input-group-addon and the textarea? I might do it on a PR after this unless you want to include it here, it's fairly easy.

screenshot from 2015-11-06 19-20-21

**Comment by kybaker** [comment]
@tbrisker In PatternFly if a user does not fill a required field or a user enters invalid data the border becomes red [1]. If the goal is to just draw the users attention and to differentiate the field visually I think the treatment above might be okay.

As far as softening the borders - The only time softening the borders is appropriate is when a field is disabled. Even when the field is disabled it is important to have enough contrast so the field does not just look like a rectangle. [1]

[1] https://www.patternfly.org/widgets/#forms

**Comment by tbrisker** [comment]
@dLobatog I have removed the color highlighting and fixed the buttons, improved the code a bit from previous commits and squashed it all together.
I am not sure why you aren't getting tooltips, they seem to work fine for me - could you please give more info on this matter?
As for the softer borders I tend to agree with @kybaker that it doesn't look very good in this situation.

**Comment by domcleal** [comment]
> I don't see any tooltips on hover

I think that's existing bug that might happen if you just added the Puppet class.  If you save the host and come back, they ought to work.

**Comment by dLobatog** [comment]
Reviewed 6 of 20 files at r1.
Review status: 6 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[app/helpers/application_helper.rb, line 64 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2kn0_SkbtJs5BPNqNh)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/helpers/application_helper.rb#L64)):</sup>
I s this change (& the equivalent on javascript) related to the issue? 


---

<sup>**[app/models/concerns/host_common.rb, line 45 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2koNRKjzW1qb-mPUJV)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/models/concerns/host_common.rb#L45)):</sup>
Under what circumstances would the other `if` fail? 


---

<sup>**[app/views/provisioning_templates/_combinations.html.erb, line 5 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2kokb9YeZWxibKg40P)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/views/provisioning_templates/_combinations.html.erb#L5)):</sup>
I'm a bit confused, sorry, why this change?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/theforeman/foreman/2858)

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -61,7 +61,7 @@ def link_to_add_fields(name, f, association, partial = nil, options = {})
     fields = f.fields_for(association, new_object, :child_index => "new_#{association}") do |builder|
       render((partial.nil? ? association.to_s.singularize + "_fields" : partial), :f => builder)
     end
-    link_to_function(name, ("add_fields(this, \"#{association}\", \"#{escape_javascript(fields)}\")").html_safe, add_html_classes(options, "btn btn-success") )
+    link_to_function(name, ("add_fields('#{options[:target]}', '#{association}', '#{escape_javascript(fields)}')").html_safe, add_html_classes(options, "btn btn-success") )
```
Comment (dLobatog) [inline app/helpers/application_helper.rb:64]: Is this change (& the equivalent on javascript) related to the issue?

File: app/views/provisioning_templates/_combinations.html.erb [hunk 3]
```diff
@@ -2,5 +2,5 @@
   <%= f.fields_for :template_combinations do |builder| %>
     <%= render 'provisioning_templates/combination', :f => builder %>
   <% end %>
-  <p><%= link_to_add_fields('+ ' + _("Add combination"), f, :template_combinations, "provisioning_templates/combination") %></p>
 <% end %>
+<p><%= link_to_add_fields('+ ' + _("Add combination"), f, :template_combinations, "provisioning_templates/combination", :target => '#template_combination') %></p>
```
Comment (dLobatog) [inline app/views/provisioning_templates/_combinations.html.erb:6]: I'm a bit confused, sorry, why this change?

File: app/models/concerns/host_common.rb [hunk 4]
```diff
@@ -41,7 +41,8 @@ module HostCommon
     def lookup_values_attributes=(lookup_values_attributes)
       lookup_values_attributes.each_value do |attribute|
         attr = attribute.dup
-        if id = attr.delete(:id)
+        id = attr.delete(:id)
+        if id.present?
```
Comment (dLobatog) [inline app/models/concerns/host_common.rb:45]: Under what circumstances would the other `if` fail?

File: app/helpers/lookup_keys_helper.rb [hunk 5]
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
...
+      value_for_key = value_hash[key.id] && value_hash[key.id][key.key]
+      value, matcher = value_for_key ? [value_for_key[:value], "#{value_for_key[:element]} (#{value_for_key[:element_name]})"] : [key.default_value, _("Default value")]
+    end
 
-  def host_key_with_diagnostic(host, value_hash, key)
-    value_for_key = value_hash[key.id] && value_hash[key.id][key.key]
-    value, matcher = value_for_key ? [value_for_key[:value], "#{value_for_key[:element]} (#{value_for_key[:element_name]})"] : [key.default_value, _("Default value")]
     original_value = key.value_before_type_cast value
-    no_value = value.nil? && key.lookup_values.find_by_match("fqdn=#{host.fqdn}")
+    overriden_value = lookup_value.try(:value)
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: When would `lookup_value` not have a `:value` method?

File: app/helpers/lookup_keys_helper.rb [hunk 6]
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
...
+    no_value = value.to_s.empty? && overriden_value.to_s.empty?
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: `value.blank? && overriden_value.blank?` should be enough, see http://apidock.com/rails/Object/blank%3F

File: app/helpers/lookup_keys_helper.rb [hunk 7]
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
     klass.class_params.override.where(:environment_classes => {:environment_id => host.environment}) + klass.lookup_keys
   end
 
-  def hostgroup_key_with_diagnostic(hostgroup, key)
-    value, origin = hostgroup.inherited_lookup_value key
-    original_value = key.value_before_type_cast value
-    diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => origin}, :data => { :placement => 'top' })
-    parameter_value_content("value_#{key.key}", original_value, :popover => diagnostic_helper)
-  end
+  def key_with_diagnostic(obj, key, index, lookup_value, value_hash = nil)
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: I find this considerably harder to read than what we had before, one method for hostgroup and another for hosts. What do we gain out of merging them if we have to constantly check what object are we using inside?

File: app/views/puppetclasses/_class_parameters.html.erb [hunk 8]
```diff
@@ -1,18 +1,48 @@
...
+      <td <%= "class=has-error" if error.present? %>>
+        <div class="input-group">
+          <%= key_with_diagnostic(obj, key, i, lookup_value, value_hash) %>
+          <span class="input-group-btn">
+            <%= fullscreen_button("$(this).parent().prev()") %>
+            <% if can_edit %>
+              <%= link_to_function(icon_text('edit'), "override_class_param(this)", :title => _("Override this value"),
+                                   :'data-tag' => 'override', :class =>"btn btn-default btn-md #{'hide' if overridden}")%>
+              <%= link_to_function(icon_text('remove'), "override_class_param(this)", :title => _("Remove this override"),
+                                   :'data-tag' => 'remove', :class =>"btn btn-default btn-md #{'hide' if !overridden}")%>
```
Comment (dLobatog) [inline app/views/puppetclasses/_class_parameters.html.erb]: `if !` -> `unless`. 
In any case, how about moving this to a helper `override_button` - including the check for `can_edit`?

File: app/views/puppetclasses/_class_parameters.html.erb [hunk 9]
```diff
@@ -1,18 +1,48 @@
...
-      <td>
-        <%= content_tag :span, trunc_with_tooltip(key.key, 20), :class => "#{"override-param" if key.overridden?(obj)}", :'data-property' => 'name' %></td>
-      <td><%= obj.class.model_name == "Host" ? host_key_with_diagnostic(obj, value_hash, key) : hostgroup_key_with_diagnostic(obj, key)%></td>
-      <td>
-          <%= link_to_function(_("override"), "override_class_param(this)", :title => _("Override this value"),
-                               :'data-tag' => 'override', :class =>"btn btn-default") if authorized_via_my_scope("host_editing", "create_params") && !key.overridden?(obj) %>
+  <% overridden = key.overridden?(obj) %>
+  <% disabled = can_edit && !overridden %>
+  <% lookup_value = obj.lookup_values.detect{|v| v.lookup_key_id == key.id} %>
+  <% puppet_default = lookup_value.try(:use_puppet_default) %>
```
Comment (dLobatog) [inline app/views/puppetclasses/_class_parameters.html.erb]: This is only used once below, can you call it directly? I think you won't even need to call `.try` since you have a check `if key.type == "PuppetclassLookupKey"` there.

File: app/views/puppetclasses/_class_parameters.html.erb [hunk 10]
```diff
@@ -1,18 +1,48 @@
...
+  <% error = lookup_value.try(:errors) %>
```
Comment (dLobatog) [inline app/views/puppetclasses/_class_parameters.html.erb]: Afaik, no need to call `.try`? `:errors` is always there on ActiveRecord models

File: app/views/puppetclasses/_class_parameters.html.erb [hunk 11]
```diff
@@ -1,18 +1,48 @@
...
+            <% if can_edit %>
+              <%= link_to_function(icon_text('edit'), "override_class_param(this)", :title => _("Override this value"),
+                                   :'data-tag' => 'override', :class =>"btn btn-default btn-md #{'hide' if overridden}")%>
+              <%= link_to_function(icon_text('remove'), "override_class_param(this)", :title => _("Remove this override"),
+                                   :'data-tag' => 'remove', :class =>"btn btn-default btn-md #{'hide' if !overridden}")%>
+            <% end %>
+          </span>
+        </div>
+        <% if error.present? %>
+          <%= content_tag(:span, error.messages[:value].try(:to_sentence), :class => "help-block") %>
```
Comment (dLobatog) [inline app/views/puppetclasses/_class_parameters.html.erb]: Why `.try`? When is messages[:value] empty or something similar?

File: app/views/puppetclasses/_class_parameters.html.erb [hunk 12]
```diff
@@ -1,18 +1,48 @@
...
+        <% end %>
+      </td>
+      <td class="ca">
+        <%= check_box(value_prefix, :use_puppet_default,
+                      :value    => lookup_value.try(:id),
```
Comment (dLobatog) [inline app/views/puppetclasses/_class_parameters.html.erb]: Ditto, `:id` is always available even if it's nil

**Comment by dLobatog** [comment]
@tbrisker @orrabin Thanks, works pretty well now. I have a few questions about the code, I'm not sure I fully understand `key_with_diagnostic` and `_class_parameters`. Please see the notes inline.

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (tbrisker) [inline app/helpers/application_helper.rb:64]: This is so that the element to add to can be specified to something other then `this` - I needed this change to make the global parameter to align nicely with the puppet class parameters. (https://github.com/theforeman/foreman/pull/2858/files#diff-b74a9e3cff7e9e6af6614aea6e50e3a5R16)

File: app/views/provisioning_templates/_combinations.html.erb (hunk 3 above)
Comment (tbrisker) [inline app/views/provisioning_templates/_combinations.html.erb:6]: This is to conform to the change made above, in ApplicationHelper#link_to_add_fields

File: app/models/concerns/host_common.rb (hunk 4 above)
Comment (tbrisker) [inline app/models/concerns/host_common.rb:45]: when id=''

File: app/helpers/lookup_keys_helper.rb (hunk 5 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: `lookup_value` is `nil` when the key is not overridden for this fqdn/hostgroup.

File: app/helpers/lookup_keys_helper.rb (hunk 7 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: This is to reduce code duplication between the two different versions, most of the code is the same for both.

File: app/views/puppetclasses/_class_parameters.html.erb (hunk 9 above)
Comment (tbrisker) [inline app/views/puppetclasses/_class_parameters.html.erb]: Moved, but still need try because `lookup value` may be `nil`.

File: app/views/puppetclasses/_class_parameters.html.erb (hunk 10 above)
Comment (tbrisker) [inline app/views/puppetclasses/_class_parameters.html.erb]: ditto

File: app/views/puppetclasses/_class_parameters.html.erb (hunk 12 above)
Comment (tbrisker) [inline app/views/puppetclasses/_class_parameters.html.erb]: ditto

File: app/views/puppetclasses/_class_parameters.html.erb (hunk 11 above)
Comment (tbrisker) [inline app/views/puppetclasses/_class_parameters.html.erb]: There is a case when cloning a hostgroup that throws an error on parameter overrides because the match is invalid before saving (and on the way, highlighting all the overridden values).
In that case, `error.messages == {:match=>["is invalid"]}`, and `error.messages[:value]` is nil.

File: app/helpers/lookup_keys_helper.rb (hunk 6 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: Just realized why we did this - the value might be `false`, in which case `blank?` will return true and `to_s.empty?` will return false. reverting this back to the way it was.

**Comment by dLobatog** [comment]
I tried to fix a bit issues with `key_with_diagnostic` and `_class_parameters` on https://github.com/tbrisker/foreman/pull/2 , I'll comment in line there

File: app/helpers/lookup_keys_helper.rb [hunk 13]
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
...
+      :inherited_value => inherited_value)
+  end
+
+  def value_matcher(obj, lookup_key)
+    if obj.class.model_name == "Host"
+      value = value_hash_cache(obj)[lookup_key.id]
+      value_for_key = value.try(:[], lookup_key.key)
+      if value_for_key.present?
+        [value_for_key[:value],
+          "#{value_for_key[:element]} (#{value_for_key[:element_name]})"]
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: Indentation's off 1 space here, rubocop's unhappy

File: app/helpers/lookup_keys_helper.rb [hunk 14]
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
...
       else
-        diagnostic_class = 'warning'
-        diagnostic_helper = popover('', _("Optional parameter without value.<br/><i>Won\'t be given to Puppet.</i> <br><br><b>Description:</b> %s") % key.description, :icon => "warning-sign")
+          [lookup_key.default_value, _("Default value")]
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: Same here, indentation is off 2 spaces

File: app/models/lookup_keys/lookup_key.rb [hunk 15]
```diff
@@ -144,6 +144,12 @@ def overridden?(host)
     host.lookup_values.any? { |lv| lv.lookup_key_id == id }
   end
 
+  def overridden_value(host)
```
Comment (dLobatog) [inline app/models/lookup_keys/lookup_key.rb]: If you are going to add this change, 3 lines above you can have the previous method work in terms of this one:

``` ruby
overridden_value(host).present?
```

File: app/helpers/lookup_keys_helper.rb [hunk 16]
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
...
+      "#{parameters_receiver}_lookup_values_attributes_#{lookup_key.id}_value",
+      lookup_value.value || inherited_value,
+      :popover => diagnostic_popover(lookup_key, matcher, inherited_value, warnings),
+      :name => "#{lookup_value_name_prefix(lookup_key.id)}[value]",
+      :disabled => !lookup_key.overridden?(obj) || lookup_value.use_puppet_default,
+      :inherited_value => inherited_value)
+  end
+
+  def value_matcher(obj, lookup_key)
+    if obj.class.model_name == "Host"
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb]: You have a method `parameters_receiver` below to check the parameters_receiver based on params rather than checking the class, both are code smells though, this needs a very deep refactor (not for this PR, but we should keep it in mind). Maybe you can move this logic to the model as I try to explain below.

File: app/helpers/lookup_keys_helper.rb [hunk 17]
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
...
+      if value_for_key.present?
+        [value_for_key[:value],
+          "#{value_for_key[:element]} (#{value_for_key[:element_name]})"]
       else
-        diagnostic_class = 'warning'
-        diagnostic_helper = popover('', _("Optional parameter without value.<br/><i>Won\'t be given to Puppet.</i> <br><br><b>Description:</b> %s") % key.description, :icon => "warning-sign")
+          [lookup_key.default_value, _("Default value")]
       end
+    else #hostgroup
+      obj.inherited_lookup_value(lookup_key)
```
Comment (dLobatog) [inline app/helpers/lookup_keys_helper.rb:104]: How about moving these to their own helpers, or the model even? 
Ideally we should have a HostForm and HostgroupForm, or  call `value_matcher(obj, lookup_key)` and it should respond from the proper helper. The only reason why that doesn't work is because we load _all_ helpers on _all_ views and there's a name conflict. 

But moving this logic to the model itself would be fine and respond well. You would just call `obj.value_matcher(lookup_key)` where needed and it responds properly depending on what `obj` is, without type checking.

**Comment by dLobatog** [comment]
@tbrisker @orrabin Aside from the class checking on `value_matcher` I see this buggy behavior:
- It clears the original value if I decide to override it and then I change my mind 

Sorry for holding this one up, but I think it's important we keep this part of the code as clear as possible since 'it needs refactoring' but it may take a long time before someone actually goes ahead and does it. Right now aside from the 2 issues above I say :+1: to the rest of the PR, thanks !

File: app/models/lookup_keys/lookup_key.rb (hunk 15 above)
Comment (tbrisker) [inline app/models/lookup_keys/lookup_key.rb]: Changed. Also changed `host` to `obj` since it can be also a host group and in the future possible other objects, and added comment on why we do it by iteration instead of by DB query.

File: app/helpers/lookup_keys_helper.rb (hunk 14 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: fixed

File: app/helpers/lookup_keys_helper.rb (hunk 13 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: fixed

File: app/helpers/lookup_keys_helper.rb (hunk 16 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: There was a mistake in the `parameters_receiver` function that when `params[:host] == {}` `.present?` would return false and cause a test to fail. Changed it to `params.has_key?` instead and changed this as well.

File: app/helpers/lookup_keys_helper.rb (hunk 17 above)
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb:104]: The reason this is done in the helper is that `value_hash_cache` is only created in the helper when it is accessed. I chose not to split this into the different helpers as you suggested as it caused the code to be more difficult to follow and understand.

**Comment by tbrisker** [comment]
@dLobatog I addressed all your comments and fixed the js bug that cleared the field.

**Comment by dLobatog** [comment]
Merged as 1e1339f5683adb133234983c28853943d573b312, thanks @tbrisker & @orrabin , nice work!
